src/pybind/mgr/dashboard/controllers/__init__.py

```diff
@@ -268,24 +294,23 @@ def __init__(self, name, metadata, wait_for=5.0, exception_handler=None):
         self.exception_handler = exception_handler
 
     def _gen_arg_map(self, func, args, kwargs):
-        # pylint: disable=deprecated-method
         arg_map = {}
-        if sys.version_info > (3, 0):  # pylint: disable=no-else-return
-            sig = inspect.signature(func)
-            arg_list = [a for a in sig.parameters]
-        else:
-            sig = getargspec(func)
-            arg_list = [a for a in sig.args]
+        params = _get_function_params(func)
 
-        for idx, arg in enumerate(arg_list):
+        args = args[1:]  # exclude self
+        for idx, param in enumerate(params):
             if idx < len(args):
-                arg_map[arg] = args[idx]
+                arg_map[param['name']] = args[idx]
```

Comment:
Too many ]s and [s. Can we create a new named tuple `FuncParam` in `_get_function_params`?
```python

from collections import namedtuple
FuncParam = namedtuple('FuncParam', 'name default required')
func_params.append(FuncParam(name=param , default=val, required=False)
```

Reply:
I could but that would require a lot of changes which are out of scope of this fix. I suggest you to create an issue for that.

Reply:
The `_get_function_params` function is used in more 6 places throughout the code.

Reply:
Exactly. That's why we need to return proper types here.

Reply:
Right now, this issue is just an inconvenience to @tspmelo . Merging this as it is would introduce further technical dept.

Reply:
(Btw, I'm not blocking this PR in github)